### PR TITLE
Add offset / path to reported errors in error serialization.

### PR DESCRIPTION
**Pull request by nimakarimipour**
This PR updates error serialization contents by adding offset of program point and the path to the containing source file where the error is reported.

It will update header of `errors.tsv` file (which each row is a reported error from NullAway) from below:
```tsv
"message_type" "message" "enc_class" "enc_member" "target_kind" "target_class" "target_method" "param" "index" "uri"
```
To:
```tsv
"message_type" "message" "enc_class" "enc_member" "offset" "uri" "target_kind" "target_class" "target_method" "target_param" "target_index" "target_uri"
```

Adding offset helps users to uniquely identify reported errors across different compilations.

#### Update
In addition to changes above, this PR also adds support for backward compatibility with `SerializationAdapters`. Version `1` is considered for all outputs as is, and version `2` applies the change described above.

To control the serialization version, use the error prone flags below
```
"-XepOpt:NullAway:SerializeFixMetadataVersion=version number",

example:
"-XepOpt:NullAway:SerializeFixMetadataVersion=2",
```

If not defined the latest version will be used.

### CODE REVIEW

**Review by msridhar — APPROVED** [review]
LGTM

**Review by lazaroclapp — COMMENTED** [review]
Nothing to change on the format change itself, but this needs bumping the serialization format version number no? Also, I think we discussed making the version of the serialization format emitted configurable in case we needed to roll back the auto-annotator and keep the NullAway version at a later point?

**Review by msridhar — CHANGES_REQUESTED** [review]
Removing approved status until backwards compat is implemented

**Comment by nimakarimipour** [comment]
@msridhar @lazaroclapp Added backward compatibility and this PR is ready for another review.

**Comment by nimakarimipour** [comment]
@msridhar @lazaroclapp FYI, It seems I can only request from __one__ reviewer.

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java:167]:
```suggestion
      // use most recent version by default
```

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java:130]: Let's create a final static constant `LATEST_VERSION` in `SerializationAdapter`, and then replace `2` with `LATEST_VERSION` here and wherever else we are using `2` to mean the latest version

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java:168]: Use `LATEST_VERSION`

**Review by msridhar — COMMENTED** [review]
Mostly looks good, just some minor comments

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java
Comment (nimakarimipour) [reply nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java]: e15a83b66d593e3e046441151312aaadeb2450fc

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java
Comment (nimakarimipour) [reply nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java]: e15a83b66d593e3e046441151312aaadeb2450fc

**Review by msridhar — APPROVED** [review]

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/adapters/SerializationAdapter.java
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/fixserialization/adapters/SerializationAdapter.java:45]:
```suggestion
   * string with each field separated by a tab.
```

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java:1624]: Why is this test being removed?

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java:77]: Why not get this from `SerializationV2Adapter.getErrorsOutputFileHeader()`? Also, shouldn't we have both a test for V1 and a test for V2, based on which fix serialization config is passed? Make sure we really maintain support for both versions/formats.

**Review by lazaroclapp — COMMENTED** [review]
I think some of my comments are redundant with Manu's, but some are still disjoint, so:

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java
Comment (nimakarimipour) [reply nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java:77]: I think we discussed before it is better to see the actual expected value instead of calling APIs in tests, but changed it anyway to call `getErrorsOutputHeader()` as it is clearer. 
6c1263c96dbd73148534d62d6d552637d43f577f

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java
Comment (nimakarimipour) [reply nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java]: This is already resolved as part of e15a83b66d593e3e046441151312aaadeb2450fc

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java
Comment (nimakarimipour) [reply nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java]: This is already resolved as part of e15a83b66d593e3e046441151312aaadeb2450fc

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java
Comment (nimakarimipour) [reply nullaway/src/main/java/com/uber/nullaway/fixserialization/FixSerializationConfig.java]: This is already resolved as part of e15a83b66d593e3e046441151312aaadeb2450fc

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java
Comment (nimakarimipour) [reply nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java:1624]: Should we still serialize the version NullAway is configured ? That seemed redundant as we are setting the version ourself in the configurations. Please let me know if you would like to have it back. Thank you.

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java
Comment (nimakarimipour) [reply nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java]: Added a test for serialization version 1. f7f8b32f31262f3bdb506aef5038876d082d6749 and b6df3e5ab4df194cd7d36010d19ba1240d88ca2c.

Currently all tests are running with serialization version `2`. With this added test, we specifically test the expected output conforms with version `1`. There might be better ways to set serialization version via `SerializationTestHelper` but I thought doing that might be out of scope of this PR and we do it in a follow up PR. Please let me know if you would like to have it in the same PR.

**Comment by nimakarimipour** [comment]
@lazaroclapp Thank you for the review, this is ready for another round. I ran it with the [caching](https://github.com/ucr-riple/NullAwayAnnotator/pull/111) PR I prepared to make sure we have everything we need before requesting a review. I noticed we need an extra column in the serialized errors which is the path to the source file where the error is reported. It is required as we need to update the offset and we need the path to the file to retrieve its latest offset changes. I just added and it is ready for review.

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java
Comment (lazaroclapp) [reply nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java:1624]: Ah, the idea is that now, instead of the auto-annotator checking which version NullAway is using for data serialization, it _asks_ for a specific version?

So, here is the problem, what happens with existing versions of the auto-annotator, which need to still figure out that they are dealing with e.g. version 1 of the format vs version 0.

Actually, if the fix serialization version is being loaded from XML configuration, rather than directly from an Error Prone option... how do I downgrade the version to configure this to work with the existing v1.3.4 of the auto-annotator, which is the entire goal of this compatibility?

Comment (nimakarimipour) [reply nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java:1624]: Brought is back as we discussed offline and we can set the serialization version via error prone flags.

a8bac9fc0f856b1cbf82ceb4b22fdf0675f69dc9

File: nullaway/src/main/java/com/uber/nullaway/fixserialization/Serializer.java
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/fixserialization/Serializer.java:130]: Why add an empty String?
```suggestion
      fileWriter.write(Integer.toString(serializationAdapter.getSerializationVersion()));
```

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java:151]: Actually, for tests that aren't explicitly about a particular serialization version, let's remove this and and have it use the latest by default. That will make it easier to detect any effect on the tests for future serialization version changes. It might still be worth having some version specific sanity tests, of course, but for the vast majority of test cases, we should just have it default to the latest.

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java:1455]: Can we make a helper method for this test that takes the desired version number and then two one-liner test cases that call it for 1 and 2?

**Review by lazaroclapp — COMMENTED** [review]
Overall LGTM, just two small nits and also I think test cases should default to the latest serialization available unless specifically meant to test one version or another (your call on whether we want to add the test case for version 2 explicitly now, or wait until there is a version 3, but the vast majority of test cases should just pick up the latest one :) )

**Review by msridhar — APPROVED** [review]
LGTM assuming @lazaroclapp's comments are addressed

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java
Comment (nimakarimipour) [reply nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java]: Sure, sounds good. Removed explicitly setting to version `2`. We can add specific unit tests for version `2` once we have version `3` to verify backward compatibility, just like how we did it for `1`.

File: nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java
Comment (nimakarimipour) [reply nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java:1430]: Sure, fad05620a6e5365c25e2eb138847e2b805c13959

**Comment by nimakarimipour** [comment]
@msridhar I think this is ready now. Please let me know if there is anything else.

**Comment by msridhar** [comment]
I see the comments are addressed; merging

**Pull request merged**
